src/tools/interop/idt/capture/platform/android/android.py

```diff
@@ -170,48 +165,50 @@ def _authorize_adb(self) -> None:
+            self.logger.info('Confirming authorization, press enter after auth')
             input('')
-        print(f'Target android device ID is authorized: {self.device_id}')
-
-    def check_screen(self) -> bool:
-        screen_cmd_output = self.run_adb_command(
-            self.check_screen_command, capture_output=True)
-        return "true" in screen_cmd_output.get_captured_output()
-
-    async def prepare_screen_recording(self) -> None:
-        if self.screen_proc.command_is_running():
-            return
-        try:
-            async with asyncio.timeout_at(asyncio.get_running_loop().time() + 20.0):
-                screen_on = self.check_screen()
-                print("Please turn the screen on so screen recording can start!")
-                while not screen_on:
-                    await asyncio.sleep(2)
-                    screen_on = self.check_screen()
-                    if not screen_on:
-                        print("Screen is still not on for recording!")
-        except TimeoutError:
-            print("WARNING screen recording timeout")
-            return
+        self.logger.info(f'Target android device ID is authorized: {self.device_id}')
+
+    async def connect(self) -> None:
+        if not self.connected:
+            self._authorize_adb()
+            self.capabilities = Capabilities(self)
+            self.capabilities.check_capabilities()
+            for stream in streams.__all__:
+                self.streams[stream] = getattr(streams, stream)(self)
+            self.connected = True
+
+    async def handle_stream_action(self, action: str) -> None:
+        had_error = False
+        for stream_name, stream in self.streams.items():
+            self.logger.info(f"Doing {action} for {stream_name}!")
+            try:
+                await getattr(stream, action)()
+            except Exception:
+                self.logger.error(traceback.format_exc())
+                had_error = True
+        if had_error:
+            raise Exception("Propagating to controller!")
 
     async def start_streaming(self) -> None:
-        await self.prepare_screen_recording()
-        if self.check_screen():
-            self.pull_screen = True
-            self.screen_proc.start_command()
-        self.logcat_proc.start_command()
-
-    async def pull_screen_recording(self) -> None:
-        if self.pull_screen:
-            self.screen_proc.stop_command()
-            print("screen proc stopped")
-            await asyncio.sleep(3)
-            self.run_adb_command(self.screen_pull_command)
-            print("screen recording pull attempted")
-            self.pull_screen = False
+        await self.handle_stream_action("start")
+
+    async def run_observers(self) -> None:
+        observer_tasks = []
+        for stream_name, stream in self.streams.items():
+            observer_tasks.append(asyncio.create_task(stream.run_observer()))
 
     async def stop_streaming(self) -> None:
```

Comment:
Should you clean up the observer tasks?

Reply:
Good catch thanks!

Reply:
Actually the controller is managing the root observer task here in `run_analyzer`.

Reply:
Here is an example placed in `screen` (not committed) to show that the root task being cancelled cancels the subtask (no more references).

```
git diff idt/capture/platform/android/streams/screen/screen.py
diff --git a/src/tools/interop/idt/capture/platform/android/streams/screen/screen.py b/src/tools/interop/idt/capture/platform/android/streams/screen/screen.py
index a18ca25f9d..3e1b97176b 100644
--- a/src/tools/interop/idt/capture/platform/android/streams/screen/screen.py
+++ b/src/tools/interop/idt/capture/platform/android/streams/screen/screen.py
@@ -79,11 +79,14 @@ class ScreenRecorder(AndroidStream):
             self.logger.info(f"New screen recording file started {self.screen_phone_out_path} {self.screen_artifact}")
 
     async def run_observer(self) -> None:
-        while True:
-            if not self.screen_proc.command_is_running():
-                self.logger.warning(f"Screen recording proc needs restart (may be normal) {self.platform.device_id}")
-                await self.start()
-            await asyncio.sleep(4)
+        try:
+            while True:
+                if not self.screen_proc.command_is_running():
+                    self.logger.warning(f"Screen recording proc needs restart (may be normal) {self.platform.device_id}")
+                    await self.start()
+                await asyncio.sleep(4)
+        except asyncio.CancelledError:
+            self.logger.critical("SCREEN RECORDER BEING CANCELLED BY CONTROLLER")
 
     async def pull_screen_recording(self) -> None:
         if self.screen_pull:
```

Output:

```
________________________
Cancelling monitor tasks
________________________

2023-11-05 19:52:11,200 INFO {controller} [run_analyzers] Done cancelling monitor tasks

_________________________
Cancelling analysis tasks
_________________________

2023-11-05 19:52:11,200 INFO {controller} [run_analyzers] Done cancelling analysis tasks
2023-11-05 19:52:11,200 INFO {play_services} [analyze_capture] Closing logcat stream
2023-11-05 19:52:11,200 CRITICAL {screen} [run_observer] SCREEN RECORDER BEING CANCELLED BY CONTROLLER
```

Reply:
I will however explicitly clean up by catching the cancel exception in android, thanks again!

Reply:
Actually I believe whats really happening here is GC... The root android task does not run continuously, rather starts stream tasks and exits. There is still a reference to the android task after it exits (we try to cancel it just in case) which has a local reference  to the stream tasks, but that reference isn't practically useful, however the stream task is still not GC until the android task is. There is no useful reason to continuously run the android analyze task on a loop at the moment, but I've done it anyway at a slow tick and to make the cancel more understandable.